sql/parser/eval.go

```diff
@@ -777,68 +901,42 @@ func evalComparisonOp(op ComparisonOp, left, right Datum) (Datum, error) {
 		}
 	}
 
-	not := false
-	switch op {
-	case NE:
-		// NE(left, right) is implemented as !EQ(left, right).
-		not = true
-		op = EQ
-	case GT:
-		// GT(left, right) is implemented as LT(right, left)
-		op = LT
-		left, right = right, left
-	case GE:
-		// GE(left, right) is implemented as LE(right, left)
-		op = LE
-		left, right = right, left
-	case NotIn:
-		// NotIn(left, right) is implemented as !IN(left, right)
-		not = true
-		op = In
-	case NotLike:
-		// NotLike(left, right) is implemented as !Like(left, right)
-		not = true
-		op = Like
-	case NotSimilarTo:
-		// NotSimilarTo(left, right) is implemented as !SimilarTo(left, right)
-		not = true
-		op = SimilarTo
-	case IsDistinctFrom:
-		// IsDistinctFrom(left, right) is implemented as !EQ(left, right)
-		//
-		// Note the special handling of NULLs and IS DISTINCT FROM above.
-		not = true
-		op = EQ
-	case IsNotDistinctFrom:
-		// IsNotDistinctFrom(left, right) is implemented as EQ(left, right)
-		//
-		// Note the special handling of NULLs and IS NOT DISTINCT FROM above.
-		op = EQ
-	case Is:
-		// Is(left, right) is implemented as EQ(left, right)
-		//
-		// Note the special handling of NULLs and IS above.
-		op = EQ
-	case IsNot:
-		// IsNot(left, right) is implemented as !EQ(left, right)
-		//
-		// Note the special handling of NULLs and IS NOT above.
-		not = true
-		op = EQ
+	// Make sure the expression's cmpOp function is memoized
+	if expr.fn.fn == nil {
+		if _, err := typeCheckComparisonExpr(expr); err != nil {
+			return DNull, err
+		}
+
+		// If cmpOp's function is still nil, return unsupported op error
+		if expr.fn.fn == nil {
+			return DNull, fmt.Errorf("unsupported comparison operator: <%s> %s <%s>",
+				left.Type(), expr.Operator, right.Type())
+		}
 	}
 
-	// TODO(pmattis): Memoize the cmpOps lookup as we've done for unaryOps and
-	// binOps.
-	if f, ok := cmpOps[cmpArgs{op, reflect.TypeOf(left), reflect.TypeOf(right)}]; ok {
-		d, err := f(left, right)
-		if err == nil && not {
-			return !d, nil
+	_, newLeft, newRight, not := foldComparisonExpr(expr.Operator, left, right)
+
+	// Initialize cache value once if needed
+	if expr.fn.initFn != nil {
+		expr.cacheInit.Do(func() {
```

Comment:
Use of `sync.Once` here is overkill as goroutine-safety is not needed. Though, if you take the path of my other comment and get rid of `initFn`, this will go away completely.